src/shared/issue.ts

```diff
@@ -1,16 +1,21 @@
 import { Context } from "../types/context";
 
+export function returnOptional<T>(value: T | undefined): T {
+  if (value === undefined) throw new Error("Value is undefined");
+  return value;
+}
```

Comment:
I've never seen this style in JS, it feels like Rust and I don't know if I like or not - it's a bit confusing when looking at the code.
@gentlementlegen what do you think?

Reply:
I'd say it is a bit strange to throw on `undefined`, I would either use a type guard directly which would make more sense in a TypeScript context, or a validator of some sort. So not against it, but might be better options.

Reply:
I felt like this was a typeguard against `undefined` so an anti-typeguard of sorts

>  use a type guard directly

If `@octokit/rest` types it as undefined all we can do is cast it, throw if it's undefined or use something like `""` or `null`

Reply:
Below does not resolve TS warnings because where `returnOptional()` has been used requires that the return val not be `undefined` but we need `T` to extend `undefined` because we are passing args in which are typed as `undefined` by octokit.

```ts
type UndefinedConstraint<T> = T extends undefined ? never : T;

function undefinedGuard<T>(value: T): value is UndefinedConstraint<T> {
  return value !== undefined;
}

export function returnOptional<T extends string | undefined = string>(value: T): T {
  if (undefinedGuard(value)) {
    return value;
  }
  throw new Error("Value is undefined");
}
```
> I'd say it is a bit strange to throw on undefined

I'd agree but if we do not stop execution we return `undefined`. 

> or a validator of some sort

I don't quite understand what you mean here, how can I validate it beyond asserting it is or is not `undefined`? 

If ever any of these values are actually undefined during runtime the octokit calls would not work anyway and so we would want to throw an error in each case?


---

In order to replace casting across the board and not using my approach, we'd need to destructure and assert these values early on and then pass them down. This means no more off-the-cuff destructing of `context.payload` as they'll need asserted or casted.

---

Typeguards focus the scope of types based on discrimination but if the underlying value is typed as `undefined` a typeguard won't change that.

We could use `Required<T>` and force all props of the type be non-optional but seems like a poor blanket approach.

We could maybe get fancy with types and remove `undefined | null` from those that we know will always be defined but I think that will look a lot more complex than what's needed for this.

---

Reply:
I'm not against throwing if a value is undefined (where it makes sense), in fact it's a right thing to do, but usually you do that with a simple if and throw. I assume you're trying to reduce repeating code by putting it in a function. However we never use that pattern in other repos and I think sometimes it's okay to be a little verbose.

Reply:
> I assume you're trying to reduce repeating code by putting it in a function

Exactly yeah it's called 9 times throughout this codebase alone

I feel this is the first time casting has been scrutinized so heavily but it's a good thing imo. So the approach to use is 'if-and-throw' and never cast unless necessary